simpleclient/src/main/java/io/prometheus/client/SimpleCollector.java

```diff
@@ -53,30 +51,155 @@
+            if (size != other.size()) {
+                return false;
+            }
+            for (int i = 0; i < size; i++) {
+                final Object a = get(i);
+                final Object b = other.get(i);
+                final boolean eq = (a == b) || (a != null && a.equals(b));
+                if (!eq) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        /**
+         * Returns the hash code value for this list.
+         *
+         * <p>This implementation uses exactly the code that is used to define the
+         * list hash function in the documentation for the {@link List#hashCode}
+         * method.
+         *
+         * @return the hash code value for this list
+         */
+        public int hashCode() {
+            int hashCode = 1;
+            for (int i = 0, size = size(); i < size; i++) {
+                final Object e = get(i);
+                final int objHash = (e == null ? 0 : e.hashCode());
+                hashCode = 31 * hashCode + objHash;
+            }
+            return hashCode;
+        }
     }
-    for (String label: labelValues) {
-      if (label == null) {
-        throw new IllegalArgumentException("Label cannot be null.");
-      }
+
+    /**
+     * Return the Child with the given labels, creating it if needed.
+     * <p>
+     * Must be passed the same number of labels are were passed to {@link #labelNames}.
+     */
+    public Child labels(String... labelValues) {
+        validateLabels(labelValues);
+        final List<String> labels;
+        if (labelValues.length > 0) {
+            labels = pooledLabelNamesOf(labelValues);
+        } else {
+            labels = Collections.emptyList();
+        }
+        return getOrCreateChild(labels);
     }
-    List<String> key = Arrays.asList(labelValues);
-    Child c = children.get(key);
-    if (c != null) {
-      return c;
+
+    private List<String> pooledLabelNamesOf(String... labelValues) {
+        final ThreadLocal<ArrayList<String>> labelNamesPool = this.labelNamesPool;
+        ArrayList<String> pooledLabels = labelNamesPool.get();
+        final int labelValuesCount = labelValues.length;
+        if (pooledLabels == null) {
+            pooledLabels = new LabelNames(labelValuesCount);
+            labelNamesPool.set(pooledLabels);
+        }
+        for (String label : labelValues) {
+            pooledLabels.add(label);
+        }
+        return pooledLabels;
+    }
+
+    private Child getOrCreateChild(List<String> labels) {
+        Child c = children.get(labels);
+        if (c != null) {
+            labels.clear();
+            return c;
+        }
+        return tryCreateChild(labels);
+    }
+
+    private Child tryCreateChild(List<String> labels) {
+        Child c2 = newChild();
+        Child tmp = children.putIfAbsent(labels, c2);
+        if (tmp == null) {
+            labelNamesPool.set(null);
```

Comment:
Why do you throwaway the pool here?

Reply:
Because if labels is being put into children is better to not sharing it to others that will come later by mean of the pool: the risk is to having the same list in `children` and pooled.

Reply:
Can you add a comment on that?

Reply:
Comment added :+1: